README.md
# parsegonio
[![DOI](https://zenodo.org/badge/122348759.svg)](https://zenodo.org/badge/latestdoi/122348759)

* A hack to parse the [Argos Goniometer](https://www.clsamerica.com/argos-goniometer) log data into a `prv` file readable by Argos Message Decoder / DAP Processor or WC portal.

* **warning:** No error handling! Further testing required! Use with caution. This is still under active dev and may change without backwards compatibility. 

* Ultimately, I hope a version of this will be included in the [sattagutils](https://github.com/williamcioffi/sattagutils) package.

* see below for a quick guide.

* Report any bugs to the [issues](https://github.com/williamcioffi/parsegonio/issues) page. 

* If you're interested in contributing, start by taking a peak at the [guidelines](CONTRIBUTING.md).

* For now, please cite as:

Cioffi WR (2020). parsegonio: parse Argos Goniometer log data and convert into prv. R code version 0.1.0. https://github.com/williamcioffi/parsegonio. (doi:10.5281/zenodo.3647697).

## Quick start

Take a look at `testparsegonio.R `which includes a full working example to output a `prv` file.

## Details

We use the Goniometer in hyperterminal mode connected to a laptop and generating plain text log files. There is a formatted output for "favorite" platforms which you get when you click save messages, but since there is a 16 platform limit I prefer to use the hyperterimal logs as `parsegonio` input. See [below](#creating-a-prv-from-the-exported-messages-of-a-favorite-platform) for instructions on using the \*.xls output of saved messages.

Goniometer log output usually looks something like this:

```
2017-05-02 14:42:07 : USB Connection to RXG134
2017-05-10 21:00:38 Received : $NPRF,7,17,5,10,21,0,24,0A1FBD4,401677726,7,109,115,-128,-128,-4486401,2133099,10,192,D4049D095018019931F7F4B00960508485440003B6567FC0*46
2017-05-10 21:00:58 Received : $NPRF,7,17,5,10,21,0,44,0A1FBD4,401677737,7,117,113,-127,-127,-4486412,2133177,2,192,D4040555502CF64C3207F4A009605084854000027FC00000*01
```

The $NPRF indicates that a platform has been favorited, but the log will also display messages from non favorited platforms with the label $NPR. The date, PTT, signal strength, bearing, among other things are recorded here. The last field is the data message.

You can use `parsegonio` to create a simulated prv file that DAP processor can read:

```r
gfile <- "gonio_ex_log.txt"
pttkey_file <- "pttkey.csv"

source("parsegonio.r")
output <- parsegonio(gfile, pttkey_file)
cat(output, file = "gonio_output.prv")
```

`parsegonio` takes two parameters: 
- `pttkey_file` is a `csv` which includes the PTT, hex, and DeployID of platforms of interest (only the PTT and hex are necessary)
- `gfile`, in this case an example Goniometer log included in this repo.

I’ve saved the output to `gonio_output.prv` here. It seems like there is some flexibility in the `prv` format and I’ve taken some liberties, mainly because I don’t entirely understand every part of the format. In my simulated `prv`, for each platform, all the messages from the Goniometer log are lumped under one satellite pass even if they occurred over many days. This isn’t realistic, but DAP Processor doesn’t seem to mind and satellite passes don’t mean anything in this use case anyway. In addition, real `prv` files include the Doppler Argos position. Obviously I don’t have one, so I just added a point near our field site. You can change add your own positions as strings with three decimal placers using the parameters `lon` and `lat`. The format appears to be 0-360 instead of -180 to +180 by the way. There is some more chitchat about all of this in [issue #4](https://github.com/williamcioffi/parsegonio/issues/4).

Once you have the simulated prv output you can just drag it into DAP processor. It helps to have preloaded a workspace with wch files as well as DAP will use the tag settings during decoding, though you can still get something even without them.

![](docs/images/dap_ex.png)

`csv` files of various datastreams can now be exported. Some of these for instance, `*-Locations.csv`, `*-Argos.csv` will be complete junk. Others will mostly make sense, but be sure to ignore anything to do with Doppler positions or satellite passes.

## One word of caution about ptts and csvs

If you are opening any `csv` files in excel and if your hex has an E somewhere in the middle and all the other digits are numbers (not letters) then excel will interpret it as scientific notation. For example `12345E2` will be converted into `12345 x 10^2` by excel. This is quite annoying and will happen every time you open the file. It’ll look something like this:

![](docs/images/badhex.png)

The best solution is don’t use excel because it is a terrible `csv` editor. But many folks are most comfortable with editing `csv` files in excel so in the past I’ve added a notes column that starts with text so the real hex can be recovered if someone accidentally edits it in excel and saves the result.

![](docs/images/savedhex.png)

## Creating a prv from the exported messages of a favorite platform
I've included an example file `favorite_ptt_180754.xls` obtained by using the save messages option for a favorited platform. To create a `prv` file, first simply save it as a csv. The expected date format is `%d-%m-%y %H:%M:%S` (this is currently hardcoded around line 186 if you have to change it). After saving a csv file you can simply run something like this and decode the prv as explained above:

```r
gfile <- "favorite_ptt_180754.csv"
ptt <- "180754"

source("parsegonio.r")
output <- parsegonio_favorite_messages(gfile, ptt)
cat(output, file = "gonio_favorite_output.prv")
```


NEWS.md
# parsegonio v0.1.1
* added a similar function named `parsegonio_favorite_messages` to parse messages that were exported using the save messages option for a favorited platform.
* user adjustable lon and lat (still one per prv file though)

# parsegono v0.1.0
* initial release for testing


parsegonio.r
# parse the raw goniometer files

parsegonio <- function(gfile, pttkey_file) {
# constants for the fake prv file
# header
h1 <- "03126"
h2 <- ""	# this is the ptt
h3 <- " 75 31 A 2"
h4 <- ""	# this is the date
h5 <- ""    # this is the time
h6 <- " 37.075  285.342  0.000 401677432"

# footer
f1 <- "" 	# this is the number of messages
f2 <- "msgs 000>-120dB  Best:  -126  Freq:  677432.3  IQ : 66" # first line
f3 <- "Lat1: 37.075N  Lon1:  74.658W  Lat2: 44.998N  Lon2:  34.063W" # second line

# read in raw goniometer file and separate the favorited NPRF from the NPR.
g <- readLines(gfile)
g_nprf <- g[grep("NPRF", g)]
g_npr <- g[grep("NPR,", g)]

# make new files out of these greped vectors
if(length(g_nprf) == 0 & length(g_npr) == 0) stop("you didn't give me any goniometer messages!")

allg <- data.frame(
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character()
)

names(allg) <- paste0("V", 1:20)

if(length(g_nprf) != 0) {
	nprf_file <- tempfile()
	writeLines(g_nprf, nprf_file)
	nprf <- read.table(nprf_file, header = FALSE, sep = ',', stringsAsFactors = FALSE)
	allg <- rbind(allg, nprf)
}

if(length(g_npr) != 0) {
	npr_file <- tempfile()
	writeLines(g_npr, npr_file)
	npr <- read.table(npr_file, header = FALSE, sep = ',', stringsAsFactors = FALSE)
	# add dummy columns for the saved average strength and average bearing columns for NPR (this only works on NPRF)
	npr_withcols <- data.frame(npr[, 1:12], rep(NA, nrow(npr)), npr[, 13], rep(NA, nrow(npr)), npr[, 14:ncol(npr)])
	names(npr_withcols) <- paste0("V", 1:20)
	allg <- rbind(allg, npr_withcols)
}

if(nrow(allg) == 0) stop("i don't think you have anything to work with there...")

# remove excess words from recieved date time
rec_date <- as.character(allg$V1)
rec_date_split <- strsplit(rec_date, split = " ")
rec_date_formatted <- sapply(rec_date_split, function(l) paste(l[1], l[2]))

allg$V1 <- rec_date_formatted

# separate the byte after the asterix
msg <- as.character(allg$V20)
msg_split <- strsplit(msg, "\\*")
msg_formatted <- sapply(msg_split, '[[', 1)
msg_asterix <- sapply(msg_split, '[[', 2)

allg$V20 <- msg_formatted
allg[, 'asterix'] <- msg_asterix

# read in the pttkey
pttkey <- read.table(pttkey_file, header = TRUE, sep = ',', stringsAsFactors = FALSE)
desehex <- pttkey$HEX[which(pttkey$DEPLOYID != "")]

# look for just those hex codes
subg <- allg[which(allg$V9 %in% desehex), ]
foundhexes <- unique(subg$V9)
foundptts <- pttkey$PTT[match(foundhexes, pttkey$HEX)]
output <- ""
	
# go through each found hex and bundle up all the messages
pb <- txtProgressBar(style = 3)
for(i in 1:length(foundhexes)) {
setTxtProgressBar(pb, i/length(foundhexes))
	dese <- which(subg$V9 == foundhexes[i])
	header <- paste(h1, foundptts[i], h3, subg$V1[dese[1]], h6)
	
	date <- subg[dese, ]$V1
	msgs <- subg[dese, ]$V20
	msgs <- as.character(msgs)
	msgs <- split(msgs, 1:length(msgs))
	
	# split up into bytes dropping first byte
	msgs <- lapply(msgs, function(l) {
		lprime <- substring(l, seq(1, nchar(l), by = 2), seq(2, nchar(l), by = 2))
		lprime[-1]
	})
	
	# convert from hex to decimal and add 1 leader zero
	msgs <- lapply(msgs, as.hexmode)
	msgs <- lapply(msgs, as.integer)
	msgs <- lapply(msgs, function(l) {
		dese <- which(nchar(l) == 1)
		l[dese] <- paste0("0", l[dese])
		l
	})
	
	output <- paste0(output, header, "\n")

	for(p in 1:length(dese)) {
		output <- paste0(output, date[p], " 1")
		curmsg <- msgs[[p]]
		msgseq <- rep(1:ceiling(length(curmsg)/4), each = 4)
		msgseq <- msgseq[1:length(curmsg)]
		u_msgseq <- unique(msgseq)
		
		for(q in 1:length(u_msgseq)) {
			dose <- which(msgseq == u_msgseq[q])
			output <- paste0(output, "\t", paste(curmsg[dose], collapse = "\t"), "\n")
		}
		
	}
}
close(pb)

if(length(output) == 0) warning("i don't think you found any matching hex...")

output
}


parsegonio.R
# parse the raw goniometer files

parsegonio <- function(gfile, pttkey_file, prv_output = TRUE, lon = '283.328', lat = '34.716') {
# constants for the fake prv file
# header
h1 <- "03126"
h2 <- ""	# this is the ptt
h3 <- " 75 31 A 2"
h4 <- ""	# this is the date
h5 <- ""    # this is the time
h6 <- paste0(" ", lat, "  ", lon, "  0.000 401677432")

# footer
f1 <- "" 	# this is the number of messages
f2 <- "msgs 000>-120dB  Best:  -126  Freq:  677432.3  IQ : 66" # first line
f3 <- "Lat1: 34.716N  Lon1:  76.672W  Lat2: 34.716N  Lon2:  76.672W" # second line

# read in raw goniometer file and separate the favorited NPRF from the NPR.
g <- readLines(gfile)
g_nprf <- g[grep("NPRF", g)]
g_npr <- g[grep("NPR,", g)]

# make new files out of these greped vectors
if(length(g_nprf) == 0 & length(g_npr) == 0) stop("you didn't give me any goniometer messages!")

allg <- data.frame(
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character(),
character()
)

names(allg) <- paste0("V", 1:20)

if(length(g_nprf) != 0) {
	nprf_file <- tempfile()
	writeLines(g_nprf, nprf_file)
	nprf <- read.table(nprf_file, header = FALSE, sep = ',', stringsAsFactors = FALSE)
	allg <- rbind(allg, nprf)
}

if(length(g_npr) != 0) {
	npr_file <- tempfile()
	writeLines(g_npr, npr_file)
	npr <- read.table(npr_file, header = FALSE, sep = ',', stringsAsFactors = FALSE)
	# add dummy columns for the saved average strength and average bearing columns for NPR (this only works on NPRF)
	npr_withcols <- data.frame(npr[, 1:12], rep(NA, nrow(npr)), npr[, 13], rep(NA, nrow(npr)), npr[, 14:ncol(npr)])
	names(npr_withcols) <- paste0("V", 1:20)
	allg <- rbind(allg, npr_withcols)
}

if(nrow(allg) == 0) stop("i don't think you have anything to work with there...")

# format date
rec_date_tmp <- paste(
	allg$V3, 				# this code might break in the year 3000?
	sprintf("%02d", allg$V4), 
	sprintf("%02d", allg$V5),
	sep = '-'
)

rec_date_tmp <- as.character(as.POSIXct(rec_date_tmp, format = "%y-%m-%d", tz = "UTC"))

rec_time_tmp <- paste(
	sprintf("%02d", allg$V6), 
	sprintf("%02d", allg$V7), 
	sprintf("%02d", allg$V8),
	sep = ":"
)

# final formatted date and time
rec_date_formatted <- paste(rec_date_tmp, rec_time_tmp)
allg$V1 <- rec_date_formatted

# save back into the dataframe
allg$V1 <- rec_date_formatted

# separate the byte after the asterix
msg <- as.character(allg$V20)
msg_split <- strsplit(msg, "\\*")
msg_formatted <- sapply(msg_split, '[[', 1)
msg_asterix <- sapply(msg_split, '[[', 2)

allg$V20 <- msg_formatted
allg[, 'asterix'] <- msg_asterix

# read in the pttkey
pttkey <- read.table(pttkey_file, header = TRUE, sep = ',', stringsAsFactors = FALSE)
desehex <- pttkey$HEX[which(pttkey$DEPLOYID != "")]

# look for just those hex codes
subg <- allg[which(allg$V9 %in% desehex), ]

if(prv_output) {
  # set up vectors to make dsa
  foundhexes <- unique(subg$V9)
  foundptts <- pttkey$PTT[match(foundhexes, pttkey$HEX)]
  output <- ""
  	
  # go through each found hex and bundle up all the messages
  pb <- txtProgressBar(style = 3)
  for(i in 1:length(foundhexes)) {
  setTxtProgressBar(pb, i/length(foundhexes))
  	dese <- which(subg$V9 == foundhexes[i])
  	header <- paste(h1, foundptts[i], h3, subg$V1[dese[1]], h6)
  	
  	date <- subg[dese, ]$V1
  	msgs <- subg[dese, ]$V20
  	msgs <- as.character(msgs)
  	msgs <- split(msgs, 1:length(msgs))
  	
  	# split up into bytes dropping first byte
  	msgs <- lapply(msgs, function(l) {
  		lprime <- substring(l, seq(1, nchar(l), by = 2), seq(2, nchar(l), by = 2))
  		lprime[-1]
  	})
  	
  	# convert from hex to decimal and add 1 leader zero
  	msgs <- lapply(msgs, as.hexmode)
  	msgs <- lapply(msgs, as.integer)
  	msgs <- lapply(msgs, function(l) {
  		dese <- which(nchar(l) == 1)
  		l[dese] <- paste0("0", l[dese])
  		l
  	})
  	
  	output <- paste0(output, header, "\n")
  
  	for(p in 1:length(dese)) {
  		output <- paste0(output, date[p], " 1")
  		curmsg <- msgs[[p]]
  		msgseq <- rep(1:ceiling(length(curmsg)/4), each = 4)
  		msgseq <- msgseq[1:length(curmsg)]
  		u_msgseq <- unique(msgseq)
  		
  		for(q in 1:length(u_msgseq)) {
  			dose <- which(msgseq == u_msgseq[q])
  			output <- paste0(output, "\t", paste(curmsg[dose], collapse = "\t"), "\n")
  		}
  		
  	}
  }
  close(pb)
} else {
  output <- allg
}

if(length(output) == 0) warning("i don't think you found any matching hex...")

output
}



# use this one for a csv saved from the .xls output
# favorited messages
parsegonio_favorite_messages <- function(gfile, ptt, lon = '283.328', lat = '34.716') {
# constants for the fake prv file
# header
h1 <- "03126"
h2 <- ""	# this is the ptt
h3 <- " 75 31 A 2"
h4 <- ""	# this is the date
h5 <- ""    # this is the time
h6 <- paste0(" ", lat, "  ", lon, "  0.000 401677432")

g <- read.table(gfile, sep = ',', header = TRUE, stringsAsFactors = FALSE)

# fix date format
dates <- as.POSIXct(g$Date, tz = 'utc', format = '%d-%m-%y %H:%M:%S')
dates <- paste(dates)

header <- paste(h1, ptt, h3, dates[length(dates)], h6)
msgs <- g$Data
msgs <- as.character(msgs)
msgs <- split(msgs, 1:length(msgs))

# split up into bytes dropping first byte
msgs <- lapply(msgs, function(l) {
  lprime <- substring(l, seq(1, nchar(l), by = 2), seq(2, nchar(l), by = 2))
  lprime[-1]
})

# convert from hex to decimal and add 1 leader zero
msgs <- lapply(msgs, as.hexmode)
msgs <- lapply(msgs, as.integer)
msgs <- lapply(msgs, function(x) {
  dese <- which(nchar(x) == 1)
  x[dese] <- paste0("0", x[dese])
  x
})

output <- paste0(header, "\n")

for(p in 1:length(dates)) {
  output <- paste0(output, dates[p], " 1")
  curmsg <- msgs[[p]]
  msgseq <- rep(1:ceiling(length(curmsg)/4), each = 4)
  msgseq <- msgseq[1:length(curmsg)]
  u_msgseq <- unique(msgseq)
    
  for(q in 1:length(u_msgseq)) {
    dose <- which(msgseq == u_msgseq[q])
    output <- paste0(output, "\t", paste(curmsg[dose], collapse = "\t"), "\n")
  }
}

output
}
